1346 create recap pray and pay

v-anne opened the pull request:
This pull request implements the backend for #1346, the RECAP pray and pay project. It incorporates a few functions that (1) enforce the quota for users for maximum document requests each day, (2) that add document requests to the Prayer model, and (3) then update document requests after they are satisfied.

This is a work in progress and is just part of the MVP. Other features such as email notifications will come later.

v-anne commented:
At least two issues right now before this is ready for review.

1. Not 100% sure which file in the settings directory to define the environment variable `ALLOWED_PRAYER_COUNT `.
2. Not sure how to write the tests yet when this PR is only for the backend and the frontend isn't ready yet.

Would appreciate @mlissner's thoughts.

mlissner commented:
The var can go in misc.py or in project/prayers.py. I think the former is better until we have three or more settings for it. 

There are whole books on your second question, but I think I'd [fakers](https://faker.readthedocs.io/en/master/) to make a user and some prayers, and then I'd make some little tests to make sure my functions are working properly. This barely needs tests, I agree, but you could check that the ranking is right and that people properly run out of prayers. You could also use [time machine](https://github.com/adamchainz/time-machine) to play with the 24 hour timer?

v-anne commented:
> There are whole books on your second question, but I think I'd [fakers](https://faker.readthedocs.io/en/master/) to make a user and some prayers, and then I'd make some little tests to make sure my functions are working properly. This barely needs tests, I agree, but you could check that the ranking is right and that people properly run out of prayers. You could also use [time machine](https://github.com/adamchainz/time-machine) to play with the 24 hour timer?

Does this require me to be able to run the Docker container and everything else locally? I had issues with it the last time I tried so it might take me some time to manage it.

mlissner commented:
You could rely on Github actions to run the tests, if you want. Might be slow to iterate, but I suppose it'd work.

v-anne commented:
I'm going to confess error, @mlissner. Not sure how to solve the issue above.

mlissner commented:
@v-anne, sorry to leave you hanging. @albertisfu, do you think you could find a minute to help Vijay get their tests working?

mlissner commented:
(Just FYI, Vijay, Alberto has a few other things in his queue before this one, sorry!)

v-anne commented:
No worries, Mike. In the interest of time, I will try to work on the front end so it can be included in this PR too.

v-anne commented:
Just pushed a rough front end, I'm going to hold off on making any substantial changes to any of this until @albertisfu has a chance to weigh in. I think this PR is 85% of the way there, if Alberto can identify what the issues are with the fakers for the Prayer and RECAPDocument models, I'll be able to fully test and then have a finalized PR.

mlissner commented:
Very cool. Mind posting a couple screenshots? I can eyeball those while we wait for him to get to this. (It's on his backlog, but it's been growing rapidly this week.)

v-anne commented:
Unfortunately, I don't have much to share in terms of screenshots just yet -- the front end I pushed was very rough because of my inability to test with synthetic data.

v-anne commented:
@mlissner, this is the test case that's failing:
https://github.com/freelawproject/courtlistener/blob/aa24ba95bcd9f89e71afc1c777c86f40e5028d6f/cl/search/tests/tests_es_recap.py#L595C3-L672C1

v-anne commented:
[This test](https://github.com/freelawproject/courtlistener/blob/aa24ba95bcd9f89e71afc1c777c86f40e5028d6f/cl/search/tests/tests_es_recap.py#L595C3-L672C1) is the one that's failing. Not sure why, as I don't think I modified code that might affect it.

mlissner commented on cl/search/factories.py:
```diff
@@ -209,20 +209,23 @@ class OpinionClusterWithParentsFactory(
+    docket_entry = SubFactory(DocketEntryFactory)
+    document_type = RECAPDocument.PACER_DOCUMENT
+    pacer_doc_id = Faker("pyint", min_value=100_000, max_value=400_000)
+    is_available = Faker("boolean", chance_of_getting_true=40)
```
I think this is why your test is failing. The `AbstractPacerDocument` class, `is_available` defaults to False. Before this change, that meant that all faked RECAPDocuments would have a false value. Now 40 of them are true, so some tests will fail.

If you need this value for your test, you can just tweak the other test to explicitly set it to False.

v-anne replied:
Sorry, Mike I'm a bit confused by your comment. You're saying the original issue is the 

`AbstractPacerDocument` class defaulting to false, as seen here:

```python3
      is_available = models.BooleanField(
          help_text="True if the item is available in RECAP",
          blank=True,
          null=True,
          default=False,
      )
```

Therefore, the code in `test_available_only_field` should explicitly set `is_available` to False, but isn't that already what it's doing?

```python3
      recap_document = RECAPDocumentFactory(
          docket_entry=entry,
          description="New File",
          document_number="1",
          is_available=False,
          page_count=5,
      )
```

P.S.: could you show me how to reference lines of code inline, like your comment does?

albertisfu replied:
Hi @v-anne,  yeah Mike is right. The issue with the `RECAPDocumentFactory` and the failing tests is that we have instances of `RECAPDocumentFactory` across our codebase that don't explicitly set the `is_available` field. For example:

https://github.com/freelawproject/courtlistener/blob/8819ac1d1d5ce111769f4fceb7a46995bd606f7a/cl/lib/test_helpers.py#L1266

In those factories, we expect `is_available` to be `False`, as defined in the `AbstractPacerDocument` class.

However, the introduction of `is_available = Faker("boolean", chance_of_getting_true=40)` is causing these factories to occasionally assign a `True` value, which leads to failing tests.

I recommend removing `is_available` from the `RECAPDocumentFactory` class and explicitly setting the `is_available` values in the test as needed. For example, you can create a loop that instantiates `RECAPDocumentFactory`, assigning `is_available` randomly within the loop.


> P.S.: could you show me how to reference lines of code inline, like your comment does?

You can add a reference in the comment during a code review on the `Files Changed` tab by simply clicking the line of code and posting a comment. You can also select multiple lines at once.
![Screenshot 2024-09-13 at 1 49 32 p m](https://github.com/user-attachments/assets/980d5353-9b38-4453-a69f-4cd5ba64a6ed)

You can also reference code by using the permalink of a specific line, as I did in the comment above.
![Screenshot 2024-09-13 at 1 48 50 p m](https://github.com/user-attachments/assets/5b729473-b04e-4874-8fd6-f88fbe9db21b)

albertisfu commented:
@v-anne Sorry for the delay in checking your questions. I'm a bit lost on how I can help you at the moment. Could you let me know if the questions above have been resolved and if there's anything else I can help you with?

v-anne commented:
Hello @albertisfu, thanks for taking a look at this. I was able to fix the issue that was tripping up the tests earlier with help from Mike and you, but I don't think this is ready to merge just yet.

There are a few issues outstanding.
1. I haven't written formal tests for the Prayer logic yet. The code to generate synthetic data is done, but no tests at this point.
2. The front end isn't in a great position right now. I wrote some rough html in `cl/favorites/templates/recap_requests.html` but I was having trouble testing locally due to issues with the Docker containers. Theoretically, this part could wait as I still need to add a button to request a document's purchase on individual docket pages, but I'm open to your thoughts. I think it would be beneficial to include this page in this PR as a stub, as users won't be able to make requests just yet without the docket page buttons.

I would appreciate your help with either of those issues, particularly if you could test the front end on your machine and see if it looks alright.

mlissner commented:
I think it's probably smart for us to get your docker issue fixed, so you can properly work on this. Would it work to create a Github discussion topic about it? If it's affecting one person, it's probably affecting many.

If you want to include the view/HTML without a way for somebody to land on that page, that's all good as far as I'm concerned. Make sure you set it to private=True though in the view so that Google doesn't somehow find it either?

v-anne commented:
> I think it's probably smart for us to get your docker issue fixed, so you can properly work on this. Would it work to create a Github discussion topic about it? If it's affecting one person, it's probably affecting many.
> 
> If you want to include the view/HTML without a way for somebody to land on that page, that's all good as far as I'm concerned. Make sure you set it to private=True though in the view so that Google doesn't somehow find it either?

I will delete the containers and set them up again. They were working earlier this week, so I assume the issue is minor. If that doesn't work, I'll ask you for help.

v-anne commented:
Started discussion #4459

albertisfu commented:
@mlissner @v-anne I've applied the following changes:

- Moved the prayer methods to `cl.favorites.utils` and made them async.
- Added tests for the prayer methods.
- Added an integration test that changes the status of related Prayer instances when a document is granted.
- Fixed the `create_prayer` method so that all tests pass.
- Made the following adjustments to the `get_top_prayers` query:
    - Only checks for `RECAPDocuments` that have a related Prayer, as it was previously checking every `RECAPDocument` in the database.
    - Returns only the fields required to display on the leaderboard and create the PACER document URL.
- Improved the leaderboard template:

The HTML now looks like this:
![Screenshot 2024-09-20 at 5 08 28 p m](https://github.com/user-attachments/assets/8f4c6add-150b-4dee-8b94-a345848f3f5a)

I'm wondering which fields we need to display here and if a link to PACER is required?

I have a question:

Is it possible for a user to pray for the same `RECAPDocument` multiple times? Or should there be only one prayer per user per document? If it's the latter, we might need to add a unique constraint to the model and tweak the `prayer_eligible` method.

Is there anything else I can help with?

mlissner commented on cl/favorites/views.py:
```diff
@@ -173,3 +174,17 @@ async def view_tags(request, username):
     )
+
+
+async def open_prayers(request: HttpRequest) -> HttpResponse:
```
Can we cache this page for at least 30s, say?

v-anne replied:
Done in [8a36f56](https://github.com/freelawproject/courtlistener/pull/4386/commits/8a36f56065f3a4587b35ddec11c3e6bcce39f178)

mlissner commented on cl/favorites/urls.py:
```diff
@@ -22,4 +23,5 @@
         name="view_tag",
     ),
     path("tags/<str:username>/", view_tags, name="tag_list"),
+    path("recap_requests/", open_prayers, name="recap_requests"),
```
We'll want this URL to be better. This is the leader board of the most wanted docs in PACER. Perhaps:

`/prayers/top/`

And we could change the function name to `top_prayers`, and the template name to `top_prayers.html` as well?

v-anne replied:
Done in #[8a36f56](https://github.com/freelawproject/courtlistener/pull/4386/commits/8a36f56065f3a4587b35ddec11c3e6bcce39f178)

mlissner reviewed (changes requested):
This looks good to me with a few little things. 

Thanks Alberto for your work and review! Your comments  all sounds great.

> Is it possible for a user to pray for the same RECAPDocument multiple times? Or should there be only one prayer per user per document? If it's the latter, we might need to add a unique constraint to the model and tweak the prayer_eligible method.

Good point. Yes, we'll want to do that. 

> I'm wondering which fields we need to display here and if a link to PACER is required?

We've got a lot of work to do on that, but I think we can launch it as an MVP now, right?

v-anne commented:
Alberto, thank you so much for your work on this. My responses are below.

> I'm wondering which fields we need to display here and if a link to PACER is required?

I think this is alright for this PR. It will need some changes but this PR currently doesn't allow users to actually add prayers from the front end yet, so it's ok if it's not perfect.

> I have a question:
> 
> Is it possible for a user to pray for the same RECAPDocument multiple times? Or should there be only one prayer per user per document? If it's the latter, we might need to add a unique constraint to the model and tweak the prayer_eligible method.
> 

I agree with Mike. It was not my attention to allow that. Accordingly, I added a constraint to the `Prayer` model and adjusted the `create_prayer` method to prevent one user requesting the same document multiple times.

> Is there anything else I can help with?

I think the last thing needed before this PR can be merged is to figure out why the test suit is failing after the migration is applied.

Again, thank you for revising the issues with my code.

mlissner commented:
Looking good! It looks like you're missing the migration file and that's why tests are failing.

v-anne commented:
I think it's good to go now.

albertisfu commented on cl/favorites/views.py:
```diff
@@ -173,3 +176,18 @@ async def view_tags(request, username):
+        "top_prayers.html",
+        {
+            "top_prayers": top_prayers,
+            "private": False,  # temporary to prevent Google indexing
```
I read in a comment that this page should be hidden from search engines?  If that's correct, I think it should be set to `True`.

v-anne replied:
My understanding is Mike wanted it to be private for the time being until the other parts of the feature are complete (allowing users to request documents to be added to the list). I'll modify the code accordingly.

albertisfu reviewed (commented):
@v-anne I reviewed your changes. They look good. I just made a small tweak to the `create_prayer` method to simplify it and added a related test.

I also added the SQL migration file for the unique constraint migration.

Finally, I have a question about whether the `open_prayers` view should be private or not.

Thanks!

albertisfu commented:
@mlissner If there are no more comments from your side, I think this could be merged. It will require applying the migration on prod.

mlissner commented:
Merged! Thank you both! I'll get it deployed in a moment, over here: https://github.com/freelawproject/courtlistener/actions/runs/11002188561

mlissner commented:
We're in business! https://www.courtlistener.com/prayers/top/

mlissner commented:
I mean, sort of!
